Shared Mixins to enable effortless Studio-editable XBlocks (OC-513)

bradenmacdonald opened the pull request:
This PR contains two very useful mixin classes available for use in XBlocks:

## StudioEditableXBlockMixin

This mixin will automatically generate a working `studio_view` form that allows content authors to edit the fields of your XBlock. To use, simply add the class to your base class list, and add a new class field called `editable_fields`, set to a tuple of the names of the fields you want your user to be able to edit.

```python
@XBlock.needs("i18n")
class ExampleBlock(StudioEditableXBlockMixin, XBlock):
    ...
    mode = String(
        display_name="Mode",
        help="Determines the behaviour of this component. Standard is recommended.",
        default='standard',
        scope=Scope.content,
        values=('standard', 'crazy')
    )
    editable_fields = ('mode', 'display_name')
```

That's all you need to do. The mixin will read the optional `display_name`, `help`, `default`, and `values` settings from the fields you mention and build the editor form as well as an AJAX save handler.

If you want to validate the data, you can override `validate_field_data(self, validation, data)` and/or ` clean_studio_edits(self, data)` - see the source code for details.

Supported field types:
* Boolean: `field_name = Boolean(display_name="Field Name")`
* Float:`field_name = Float(display_name="Field Name")`
* Integer: `field_name = Integer(display_name="Field Name")`
* String: `field_name = String(display_name="Field Name")`
* String (multiline): `field_name = String(multiline_editor=True, resettable_editor=False)`
* String (html): `field_name = String(multiline_editor='html', resettable_editor=False)`
* List (set): `field_name = List(list_style='set', list_values_provider=some_method, default=[])`
  - So far a special editor is only implemented for unordered lists of unique values chosen from a small number of options. For any other type of `List`, the editor UI will be a JSON textarea.
  - The `List` declaration must include the property `list_style='set'` to indicate this.
  - The `List` declaration must also define a `list_values_provider` method which will be called with the block as its only parameter and which must return a list of possible values.
* Any of the above will use a dropdown menu if they have a pre-defined list of possible values.
* Rudimentary support for List, Dict, and any other JSONField-derived field types
  - `list_field = List(display_name="Normal List", default=[])`
  - `dict_field = Dict(display_name="Normal Dict", default={})`

Supported field options (all field types):
* `values` can define a list of possible options, changing the UI element to a select box. Values can be set to any of the formats [defined in the XBlock source code](https://github.com/edx/XBlock/blob/master/xblock/fields.py):
  - A finite set of elements: `[1, 2, 3]`
  - A finite set of elements where the display names differ from the values:
    ```
            [
             {"display_name": "Always", "value": "always"},
             {"display_name": "Past Due", "value": "past_due"},
            ]
    ```
  - A range for floating point numbers with specific increments: `{"min": 0 , "max": 10, "step": .1}`
  - A callable that returns one of the above. (Note: the callable does *not* get passed the XBlock instance or runtime, so it cannot be a normal member function)
* `resettable_editor` - defaults to `True`. Set `False` to hide the "Reset" button used to return a field to its default value by removing the field's value from the XBlock instance.

Changes compared to the existing XModule editing support in Studio:
* Editable fields are defined in a simple string-based whitelist (`editable_fields`) rather than blacklist (`non_editable_metadata_fields`)
* Can distinguish between a String field value explicitly set to `""` and a field using the default value of `""`. Studio's XModule field editor cannot.
* Can share field data validation logic with the XBlock's existing validation code - reduces code duplication
* Nicer editor for `List` fields that are really sets (unordered, unique items)

Basic screenshot:
![Basic screenshot](https://cloud.githubusercontent.com/assets/945577/6341782/7d237966-bb83-11e4-9344-faa647056999.png)

Screenshot of the list/set editor:
![Screenshot of the list or set editor](https://cloud.githubusercontent.com/assets/945577/6404805/f03c91b8-bdd0-11e4-8310-5b7090fa8a6a.png)


## StudioContainerXBlockMixin

This mixin helps with creating XBlocks that want to allow content authors to add/remove/reorder child blocks. By removing any existing `author_view` and adding this mixin, you'll get editable, re-orderable, deletable child support in Studio. To enable authors to add new children, simply override `author_edit_view` and set `can_add=True` when calling `render_children` - see the source code. To enable authors to add only a limited subset of children requires custom HTML.

An example is the mentoring XBlock:
![screen shot 2015-02-23 at 5 44 05 pm](https://cloud.githubusercontent.com/assets/945577/6341803/d0195ec4-bb83-11e4-82f6-8052c9f70690.png)


## Other changes
* Added `child_isinstance()` helper method, which uses only the public XBlock API to test a block's class given a usage_id, without the expensive operation of instantiating the block.
  - Despite the slightly convoluted code, with an edX split modulestore runtime this is at least [two orders of magnitude faster](https://gist.github.com/bradenmacdonald/0305ff697d09a653673b) than instantiating the block before checking the instance.
* Sped up `SeleniumBaseTest` by removing the initial step of loading the Workbench home page
* Added a new class for Selenium/bok_choy tests, `SeleniumXBlockTest`. It is faster and more flexible than `SeleniumBaseTest ` because it doesn't need to load a whole folder of XML scenarios. `SeleniumBaseTest ` is kept to preserve compatibility for XBlock tests that currently use it.

chrisndodge commented:
Hi @bradenmacdonald I just noticed you referencing this PR in the open edx mailing list. Do you need reviewer support on this, or is Open Craft managing this PR?

bradenmacdonald commented:
Hi @chrisndodge It's not done yet so it's still in the internal review stages. I believe we were just going to manage this PR for now (as part of my ongoing work on refactoring the mentoring XBlock), but are planning to get review from edX as part of our upcoming work to merge all the various versions of the "mentoring" XBlock and updating the version that's on edx.org. Sarina is helping us schedule that.

In the medium term, I would personally like to see the work here (or equivalent functionality implemented differently) merged into edx-platform once we've worked out the kinks by using it from xblock-utils.

However, if you have any feedback in the meantime, feel free!

CC @antoviaque who may have some corrections to what I just said :p

antoviaque commented:
@chrisndodge Thanks for asking, that's nice of you!

@bradenmacdonald Yup, that's correct - looping in @sarina so she can comment too as needed, as this will be part of the conversion of mentoring to real children effort that we have started for Harvard (cf https://github.com/open-craft/xblock-mentoring/pull/4 ), which is up for a preliminary review. I'm also currently discussing with Steve to see if we use the opportunity to also do this conversion for the edx-solution fork at the same time.

@sarina Were you able to figure out who will end up reviewing the work? Would the additional reviewer support be helpful?

sarina commented:
Xavier I sent you an email last night with some Qs if you can please
respond to that
On Feb 18, 2015 4:52 AM, "Xavier Antoviaque" <notifications@github.com>
wrote:

> @chrisndodge <https://github.com/chrisndodge> Thanks for asking, that's
> nice of you!
>
> @bradenmacdonald <https://github.com/bradenmacdonald> Yup, that's correct
> - looping in @sarina <https://github.com/sarina> so she can comment too
> as needed, as this will be part of the conversion of mentoring to real
> children effort that we have started for Harvard (cf
> open-craft/xblock-mentoring#4
> <https://github.com/open-craft/xblock-mentoring/pull/4> ), which is up
> for a preliminary review. I'm also currently discussing with Steve to see
> if we use the opportunity to also do this conversion for the edx-solution
> fork at the same time.
>
> @sarina <https://github.com/sarina> Were you able to figure out who will
> end up reviewing the work? Would the additional reviewer support be helpful?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx-solutions/xblock-utils/pull/5#issuecomment-74838333>
> .
>

antoviaque commented:
@sarina Yup, I saw it - replied!

sarina commented:
@antoviaque - I'm not yet sure who will be able to review this work. If @chrisndodge is available to do some review that would absolutely help.

chrisndodge commented:
Hi @sarina after a quick glance, I'm not sure I'm the right domain expert for these changes.

sarina commented:
@chrisndodge @antoviaque - since this is in the edx-solutions repo it falls outside of the normal OSPR process. So, not sure how you want me to help / proceed here.

antoviaque commented:
@sarina This is on an edx-solutions repository, but the change from this PR doesn't come from solutions, it was written to support the switch of mentoring to real children, ie it could be seen as a contribution from Harvard to the current repo. We could merge it with an internal code review like we normally do for edx-solutions PRs, but since this code will end up being used by the mentoring XBlock on edx.org, I am guessing that it would not be enough for edx.org, no? Ie this code would end up needing review during the mentoring update review for edx.org?

sarina commented:
@antoviaque yeah I just don't really know what the right call is, if it goes through OSPR you're going to have to provide some context to reviewers because we have no idea what this repo is. It would be great if you could get one reviewer from the Solutions team, who understands this repo and its purpose, etc, and I can provide a second technical reviewer.

sarina commented:
does this repo have tests?

antoviaque commented:
@sarina Just replied to my email suggesting exactly that : ) Sounds good, we'll do one of the reviews.

This repo has test yes (cf https://github.com/edx-solutions/xblock-utils#xblock-utils-various-utilities-for-xblocks for how to run them), but not the current PR yet, @bradenmacdonald is working on it this week to finalize the PR.

bradenmacdonald commented:
@sarina @antoviaque FYI, although this repo does have tests, all the new code in this PR depends heavily on interactions with Studio. So in the long term, I'd like to move it to edx-platform which is the only environment where we could actually test it effectively. It's here for now so that we can use the new mentoring XBlock on the Birch release (not sure how important that is), and because I assume we do not have time for the extra work and review steps required for putting it into platform.

sarina commented:
@bradenmacdonald  Hmm that's an interesting question. In the interest of getting it in, having it here makes sense, and then migrating to edx-platform later. My worry is if in the move to platform changes are requested that aren't backwards compatible with the xblock requiring a lot of stupid extra work on your part.

@antoviaque let me know if I should make an ospr ticket for this pr, or what

bradenmacdonald commented:
@Kelketek Now that I've added a bunch of integration tests, this is ready for your review. I'd like to get it merged ASAP.

n.b. To test in Studio (obviously important for this PR), you'll need to check out https://github.com/open-craft/xblock-mentoring/pull/6 and create some of those mentoring blocks in Studio. That mentoring PR is not *quite* ready for review but the parts that you need to test studio editing are done.

CC @antoviaque

antoviaque commented:
@sarina Created an OSPR ticket for the mentoring review, including the current PR: https://openedx.atlassian.net/browse/OSPR-419

@bradenmacdonald Cool - note that since this will also be part of the upstream review, it could make sense to leave the PR unmerged, and simply point the requirement in mentoring to the latest commit? This way if changes are needed they can also be included here.

antoviaque commented:
@bradenmacdonald @sarina FYI, the xblock-utils repo is now on the edx org (note that this changed the URL for this PR).

Kelketek commented on xblockutils/public/studio_edit.js:
```diff
@@ -0,0 +1,113 @@
+            dataType: "json",
+            global: false,  // Disable Studio's error handling that conflicts with studio's notify('save') and notify('cancel') :-/
+            success: function(response) { runtime.notify('save', {state: 'end'}); }
+        }).fail(function(jqXHR) {
```
I'm very interested to know if someone from upstream can comment on this-- The limitations here cause some pretty painful code duplication.

Kelketek commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,263 @@
+# Copyright (C) 2015 OpenCraft
+# License: AGPLv3
+"""
+This module contains a mixin that allows third aprty XBlocks to be easily edited within edX
```
Typo: 'aprty' should be 'party'.

Kelketek commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,263 @@
+
+class StudioEditableXBlockMixin(object):
+    """
+    An XBlock mixin to provide convenient use of an XBlock in Studio.
```
People can already use XBlocks in studio. :) Perhaps you mean convenient editing of XBlocks in Studio?

Kelketek commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,263 @@
+        # Build a list of all the fields that can be edited:
+        for field_name in self.editable_fields:
+            field = self.fields[field_name]
+            assert field.scope in (Scope.content, Scope.settings)
```
Could we add in some sort of informative error message here? Scopes are one of the most confusing things to new developers on XBlocks.

Kelketek commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,263 @@
+                        info['type'] = 'text'
+                break
+        if "type" not in info:
+            return None  # This is not a field we can support (it doesn't even have a JSON representation)
```
Can we raise a warning instead of failing silently? Or perhaps this should be a proper failure-- We're using a whitelist, after all, so the user would have had to have deliberately specified a field that doesn't work, and they'll want to see something telling them so rather than hunting down in the code to find out why the field they specified doesn't work.

Kelketek commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,263 @@
+                # e.g. {"min": 0 , "max": 10, "step": .1}
+                info["min"] = field.values["min"]
+                info["max"] = field.values["max"]
+                info["step"] = field.values["step"]
```
Is there anything in field.values that shouldn't be in info? If not, a simple info.update(field.values) may suffice here.

bradenmacdonald replied:
No, there shouldn't be. But if for some reason a key like "default" were to be set in field.values, it would mess up our info dict if we are using info.update(field.values). I think I prefer the more explicit approach in this case, so that it's clear (A) what's going on here and (B) what fields are expected/allowed.

sarina replied:
Why not be a bit safer here - perhaps doing `field.values.get("step", None)` then explicitly checking and failing for any found None values?

Kelketek commented on xblockutils/studio_editable_test.py:
```diff
@@ -0,0 +1,54 @@
+from xblockutils.base_test import SeleniumXBlockTest
+
+
+class StudioEditableBaseTest(SeleniumXBlockTest):
```
I believe that XBlock or the SDK now has bok_choy available and this is the preferable method of testing now. It's likely to come up that we're using Selenium directly here.

bradenmacdonald replied:
Good point, but if you follow the chain of inheritance, this is actually using bok-choy. https://github.com/edx/xblock-sdk/blob/master/workbench/test/selenium_test.py

Kelketek commented on README.md:
```diff
@@ -7,3 +7,7 @@ test base classes and documentation shared by many XBlocks.
 To test the utilities, run:
 
     nosetests --with-coverage --cover-package xblockutils --cover-html
```
Does this work anymore? I'm going to give it a try when I start testing, but I don't know the last time I ran this command.

bradenmacdonald replied:
No, it wasn't working for me at all. I think because it doesn't have the Django environment configured properly. That's why I added the alternate command below, which works fine. I left the old one in because I wasn't sure what I was missing, and the alternate I've suggested doesn't generate a coverage report.

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fxblock-utils&number=5

bradenmacdonald commented:
@Kelketek I responded to all review comments (in cdc6031e981f5ff299b535698df6e8159beff756) and also updated the README (d02bbf86b5ba79ae32c4ef084301f7816bda5fc5).

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-422](https://openedx.atlassian.net/browse/OSPR-422) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've added yourself to the [AUTHORS](https://github.com/edx/xblock-utils/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

cgoldberg commented on xblockutils/studio_editable_test.py:
```diff
@@ -0,0 +1,54 @@
+        return self.browser.execute_script('return window.notifications.shift();')
+
+    def wait_for_runtime_notification(self):
+        wait = WebDriverWait(self.browser, self.timeout)
```
@bradenmacdonald can the wait/synchronization be done through bok_choy?  (instead of interacting with selenium directly)
http://bok-choy.readthedocs.org/en/latest/api_reference.html#bok_choy.page_object.PageObject.wait_for

bradenmacdonald replied:
@cgoldberg Yes, I could re-write `StudioEditableBaseTest` as a bok_choy `PageObject`. Though I think for this particular method that you commented on, there is no way to execute JavaScript using pure bok choy APIs (AFAIK?) so some Selenium must be used directly anyways. I don't think I'll be able to get to that change this week though.

cgoldberg replied:
@bradenmacdonald a bok-choy PageObject has a reference to the underlying selenium webdriver instance (self.browser).  So you can execute script through that without the need to import selenium

sarina replied:
@cgoldberg personally I wouldn't consider this to be a blocker to merge, though - these tests won't be running in edx-platform main repo.

antoviaque commented:
@bradenmacdonald The inclusion of static files seem to be missing in setup.py - if this isn't installed in development mode the static files are missing from the package.

sarina commented:
@bradenmacdonald @antoviaque : in talking with Leslie, she flagged the potential similarities between this work and https://github.com/edx/edx-platform/pull/6028 - can you take a look over that pull request and let me know how it might be similar/different from this work? Christina has spent a lot of time thinking about the Studio XBlock editing interface, and we want to make sure we're not having a bunch of competing approaches in this arena.

If you have questions for Christina, please ping her on that PR.

bradenmacdonald commented:
@antoviaque Missing static files should be fixed now.

@sarina We had touched base before the recent hackathon but thanks for the reminder. I've pinger her on that PR.

hosamshahin commented:
@bradenmacdonald 

I'm trying to modify studio_view to add some logic needed by my XBlock like show or hide some fields based on the value of other dropdown field. To do that I tried to write a decorator for studio_view as following

def odsa_studio_view(func):
    def func_wrapper(*args, **kwargs):
        fragment = func(*args, **kwargs)
        fragment.add_javascript(loader.load_unicode('public/js/src/studio_edit_decorator.js'))
        fragment.initialize_js('StudioEditableXBlockDecorator')
        return fragment

    return func_wrapper

however it seems calling initialize_js for the second time causes the first function StudioEditableXBlockMixin to not to be called by the runtime. Do you have a better solution for that?

bradenmacdonald commented:
@hosamshahin Yes, that's right - you can include many JavaScript files but there is only ever one function registered at a time via initialize_js to be called when the XBlock is loaded.

What you need to do is modify your custom StudioEditableXBlockDecorator JavaScript function to call the default one. Something like this:

```javascript
function StudioEditableXBlockDecorator(runtime, element) {
    StudioEditableXBlockMixin(runtime, element); // Call the default JS code since it's no longer registered via initialize_js
    // your code here
}
```

Also, by the way, you don't need to made your studio view a decorator - a simple function override like the following is likely just as good but easier to understand.

```python
def studio_view(self, context):
    fragment = super(MyXBlock, self).studio_view(context)
    fragment.add_javascript(loader.load_unicode('public/js/src/studio_edit_decorator.js'))
    fragment.initialize_js('StudioEditableXBlockDecorator')
    return fragment
```

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+
+.. code:: python
+
+    @XBlock.needs("i18n")
```
is this sufficiently generic? do all xblocks apply this decorator?

bradenmacdonald replied:
`needs("i18n")` is a very common, useful decorator, since it's basically the only way for an XBlock to localize its translatable strings. It's not strictly required, but it's in this example code as a reminder of good practices.

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+  ``field_name = String(multiline_editor=True, resettable_editor=False)``
+* String (html):
+  ``field_name = String(multiline_editor='html', resettable_editor=False)``
+* Any of the above will use a dropdown menu if they have a pre-defined
```
This is a callout to the 6 bullets above, yet doesn't visually stand out from it. I might recommend breaking your list ie

```
* String (html):
...

Any of the above...

* List of unordered...
```

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+  ``field_name = String(multiline_editor='html', resettable_editor=False)``
+* Any of the above will use a dropdown menu if they have a pre-defined
+  list of possible values.
+* List of undordered unique values (i.e. sets) drawn from a small set of
```
typo undordered

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+    return a list of possible values.
+* Rudimentary support for List, Dict, and any other JSONField-derived field types
+
+  - ``list_field = List(display_name="Normal List", default=[])``
```
It's a bit unclear how this differs from the use of "List" in the above example. Perhaps at L77 instead of just saying "List", you can say "ordered List" or someting that distinguishes it from the list/"set" example previously.

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+
+    from xblockutils.studio_editable import StudioContainerXBlockMixin
+
+This mixin helps with creating XBlocks that want to allow content
```
This sentence is confusingly worded. Try:

This mixin helps to create XBlocks that allow content authors to add, remove, or reorder child blocks.

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+This mixin helps with creating XBlocks that want to allow content
+authors to add/remove/reorder child blocks. By removing any existing
+``author_view`` and adding this mixin, you'll get editable,
+re-orderable, deletable child support in Studio. To enable authors to
```
re-orderable, and deletable

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+authors to add/remove/reorder child blocks. By removing any existing
+``author_view`` and adding this mixin, you'll get editable,
+re-orderable, deletable child support in Studio. To enable authors to
+add new children, simply override ``author_edit_view`` and set
```
s/add new/add an unlimited number of new/

bradenmacdonald replied:
Ok, I can see this needs re-wording. These sentences were more about whether or not we want to allow authors to add arbitrary *types* of child XBlocks (and not about allowing or denying an arbitrary *number* of XBlocks). I've reworded it in my upcoming commit.

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+re-orderable, deletable child support in Studio. To enable authors to
+add new children, simply override ``author_edit_view`` and set
+``can_add=True`` when calling ``render_children`` - see the source code.
+To enable authors to add only a limited subset of children requires
```
s/subset/number/

sarina commented on README.rst:
```diff
@@ -10,4 +10,174 @@ test base classes and documentation shared by many XBlocks.
+
+If your XBlock needs to find children/descendants of a particular
+class/mixin, you should use
+``child_isinstance(self, child_usage_id, SomeXBlockClassOrMixin)``
```
Please make this call and the next call separated out for ease of reading, eg

```
you should use

.. code:: python
    child_isinstance(self, child_usage_id, SomeXBlockClassOrMixin)

rather than calling

.. code:: python
    isinstance(self.runtime.get_block(child_usage_id), SomeXBlockClassOrMixin)
```

As-is, it's hard to read the rendered RST with all the code squashed together.

sarina commented on run_tests.py:
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Run tests for XBlock-utils
```
Is it possible for you to pull out the test infrastructure into a separate PR, and set up Travis in that PR? Then you can rebase this over that and we can just add test & travis infrastructure in a logical pr.

There's some basic Travis setup here: https://github.com/edx/i18n-tools that you can probably just pull direct from.

sarina replied:
The more I think about it the more I'd like to see this - you've got a lot of tests here and it's hard to tell if they're passing or not - relying on running them locally isn't a good idea. If you open up this PR and tag me, we can get it reviewed & merged quickly.

sarina commented on tests/integration/test_studio_editable.py:
```diff
@@ -0,0 +1,299 @@
+
+    def validate_field_data(self, validation, data):
+        """ Basic validation method for these tests """
+        if data.count <=0:
```
Are you guys using any linters on this repo? This is a pep8 violation (space between operator & number)

sarina commented on tests/integration/test_studio_editable.py:
```diff
@@ -0,0 +1,299 @@
+    editable_fields = ('color', 'count', 'comment')
+
+    def validate_field_data(self, validation, data):
+        """ Basic validation method for these tests """
```
This isn't descriptive of what you're doing.

sarina commented on tests/integration/test_studio_editable.py:
```diff
@@ -0,0 +1,299 @@
+
+        comment_control = self.get_element_for_field('comment')
+        comment_control.send_keys('forcing a change')
+        comment_control.clear()
```
Why do you call the clear() after the send_keys() here, but before it in the previous two stanza?

I also wonder if there's a way to do these three (get element, clear, send_keys) in a loop over EditableXBlock.editable_fields (maybe you can send the default value or something?) so that this test is more robust to adding of any new editable fields.

bradenmacdonald replied:
Because the previous two stanzas are trying to overwrite the existing value of the field with a new value. This stanza is just trying to mark an already blank field as being modified. The difference results from the fact that two fields have a default value but the third does not.

It would be possible to write this as a loop but since each of these three fields is treated subtly differently, the loop code would be ugly and/or would not test the same range of code paths. (I want to test two different field types - string and integer - as well as a string field with a blank string as a default value that we are replacing with an explicitly set blank string)

sarina commented on tests/integration/test_studio_editable.py:
```diff
@@ -0,0 +1,299 @@
+        """
+        block = self.load_root_xblock()
+        for field_name in EditableXBlock.editable_fields:
+            self.assertFalse(block.fields[field_name].is_set_on(block))
```
these 3 lines are present in 2 tests so far - maybe they should be part of a helper method?

sarina replied:
yeah you basically do this in the assert_unchanged method at L126, maybe assert_unchanged should return the block, and be a method available to all tests.

sarina commented on tests/integration/test_studio_editable.py:
```diff
@@ -0,0 +1,299 @@
+        count_control.send_keys('10')
+
+        self.click_save(expect_success=False)
+        expect_error_message("No swearing allowed")
```
I don't understand why you expect to get this error message at this point - you've cleared the count_control and you're only sending `'10'`

bradenmacdonald replied:
The count control is now good. But this error is on the `comment` control, which is still invalid as we haven't changed it since the last time we tried clicking "save".

sarina commented on xblockutils/base_test.py:
```diff
@@ -18,27 +18,88 @@
+    """
+    Selenium Base Test for loading a whole folder of XML scenarios and then running tests.
+    This is kept for compatibility, but it is recommended that SeleniumXBlockTest be used
+    instead, since it is faster and more flexible (specifically, senarios are only loaded
```
typo: scenarios

sarina commented on xblockutils/helpers.py:
```diff
@@ -0,0 +1,17 @@
+def child_isinstance(block, child_id, block_class_or_mixin):
+    """
+    Is "block"'s child identified by usage_id "child_id" an instance of
+    "block_class_or_mixin"?
```
This could be written better. See https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015 OpenCraft
```
Other files, eg `xblockutils/base_test.py` are putting (C) edX; in edx-platform, we don't add a copyright notice at all. I'm not sure what the right thing to do here is - @antoviaque do you know?

antoviaque replied:
@sarina This comes from XBlocks where there is no requirement to assign the copyright, but it's true that since this is contributed to a repository on the edx organization, it probably should be assigned to edX and have (c) edX everywhere. 

Not sure why there isn't a copyright notice on the edx-platform though, as this is a standard good practice - maybe nobody got around to do it there when the code was released? @nedbat do you know if there is another reason?

sarina replied:
I'm fairly certain it's mostly that we've never made a push to do it.

sarina commented:
@bradenmacdonald do you have a test/toy XBlock that can be used to test out the mixin functionality? Or some good instructions? I'd like to play around with it in Studio

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+    """
+    A helper class whose attribute values come from the specified dictionary or fallback object.
+    """
+    def __init__(self, new_fields_dict, newly_removed_fields, fallback_obj):
```
I'd like to see a docstring which explains what these three parameters represent

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+            if info["default"] is None:
+                info["default"] = []
+            info["default"] = json.dumps(info["default"])
+        if info["type"] == "generic":
```
why not elif here?

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+            else:
+                # e.g. [1, 2, 3] or [ {"display_name": "Always", "value": "always"}, {...}, ... ]
+                if not isinstance(values[0], dict) or "display_name" not in values[0]:
+                    values = [{"display_name": val, "value": val} for val in values]
```
I don't understand what this is doing, given your comment at L144. If you're saying that a possibility for `values` is `[1,2,3]`, than this results in `values = [{"display_name": 1, "value": 2}]` which cannot possibly be correct.

bradenmacdonald replied:
This line is supposed to convert `[1, 2, 3]` to `[{"display_name": "1", "value": 1}, {"display_name": 2, "value": 2}, {"display_name": "3", "value": 3}]`

It does work, though it looks like I forgot to force "display_name" to string type (I will fix but it wasn't an issue since the template does that anyways).

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+                info['values'] = values
+        if info["type"] in ("list", "set") and field.runtime_options.get('list_values_provider'):
+            list_values = field.runtime_options['list_values_provider'](self)
+            # list_values must be a list of values or {"display_name": x, "value": y} objects
```
I'm wary of statements like this - if incoming data must have a particular form, I'd rather see explicit lines that asserts the form of the data and fails hard if the data's form is incorrect. That makes understanding errors _much_ easier to catch.

bradenmacdonald replied:
This code would fail hard on invalid data as is, but the errors would have been generic KeyError or similar. Will clean that up a bit.

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+                    values[field_name] = field.from_json(data['values'][field_name])
+                else:
+                    raise JsonHandlerError(400, "Unsupported field type: {}".format(field_name))
+            elif field_name in data['defaults'] and field.is_set_on(self):
```
else? should something fail here if the if and elif clauses are not hit?

bradenmacdonald replied:
Not necessarily. While it's true that the accompanying JavaScript does always send either an explicit value or a default value for every field, there's no reason to require that every editable field is always edited at the same time in our AJAX handler here. Subclasses or other potential uses of this AJAX method may need a way to say "make no change to field x", and they could do so by leaving that field out.

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+        # Example:
+        # if "name" in data:
+        #     data["name"] = data["name"].strip()
+        pass
```
is it better to pass or raise NotImplementedError here?

bradenmacdonald replied:
No, because many subclasses may not have any cleaning to do. I want this to be an optional override and not a mandatory one, in order to reduce code clutter.

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+        # Example:
+        # if data.count <=0:
+        #     validation.add(ValidationMessage(ValidationMessage.ERROR, u"Invalid count"))
+        pass
```
is it better to pass or raise NotImplementedError here?

bradenmacdonald replied:
Same thing - some subclasses may not have any validation to do, or may have more complex validation requirements that require them to overload `validate` rather than this more simplified method.

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,295 @@
+        otherwise just show the normal 'author_preview_view' or 'student_view' preview.
+        """
+        root_xblock = context.get('root_xblock')
+        is_root = root_xblock and root_xblock.location == self.location
```
Nit: I'm unclear why you need to define this variable if you're only using it in the next line. I'd err on the side of just doing, `if root_xblock and root_xblock.location == self.location:` - it's easier to read.

sarina commented on xblockutils/templates/studio_edit.html:
```diff
@@ -0,0 +1,112 @@
+                      </label>
+                    </li>
+                  {% empty %}
+                    <li>{% trans "No choices available." %}</li>
```
For this and any other text strings, they'll be translated if a translation is available in edx-platform. So I would recommend only using text that appears elsewhere in edx-platform if you actually want it translated. "No choices available" doesn't appear in edx-platform, perhaps a phrase that does appear can be used?

bradenmacdonald replied:
Ok, how about "None Available" ?

sarina commented on xblockutils/templates/studio_edit.html:
```diff
@@ -0,0 +1,112 @@
+            {% endif %}
+
+            {% if field.allow_reset %}
+              <button class="action setting-clear {% if field.is_set %}active{%else%}inactive{% endif %}" type="button" name="setting-clear" value="{% trans "Clear" %}" data-tooltip="{% trans "Clear" %}">
```
"Clear" and "Clear Value" both appear as translated strings in edx-platform (note capitalization - you have this correct)

bradenmacdonald replied:
Great, thanks for checking :)

sarina commented on xblockutils/templates/studio_edit.html:
```diff
@@ -0,0 +1,112 @@
+      </li>
+
+      <li class="action-item">
+        <a href="#" class="button cancel-button">{% trans "Cancel" %}</a>
```
"Save" and "Cancel" (both capitalized) do appear in edx-platform

sarina commented:
This is looking good. I still need to manually test it, then I'll ask someone on the platform team to take a quick look over it.

sarina commented on xblockutils/helpers.py:
```diff
@@ -0,0 +1,17 @@
+"""
+
+
+def child_isinstance(block, child_id, block_class_or_mixin):
```
child_isinstance() is an oddly specific helper method. it's not worth holding up this PR over, but i wonder if it's asking to be something more generic. it reminds me a little of the usage_key_filter param we have in get_children() in edx-platform:

https://github.com/edx/edx-platform/blob/7338cc1286783256ee3d6b7382c9460c87d1f062/common/lib/xmodule/xmodule/x_module.py#L406

bradenmacdonald replied:
You're right that it does seem oddly specific, but I feel that this exact need is not uncommon and the current definition of this helper fits flexibly into many different use cases. I worry that any attempt to make it more general would make it slower or more complex, which would defeat the purpose. For example, in all the uses I've had for it so far (in mentoring), I don't want a tree - only immediate children.

You are right that moving it into the runtime would create more avenues for optimization, but I think we need to keep the number of methods that runtimes are expected to implement to a minimum, as there are already at least several different runtimes a typical new XBlock may need to work with (workbench, LMS, Studio preview, split DescriptorSystem, old mongo DescriptorSystem).

All that being said, if we feel like this isn't going to be used much I could move it back into `mentoring` for the time being.

sarina replied:
It's not unreasonable, just something perhaps to keep in the back of your mind as the codebase evolves.

sarina commented:
Does this work on both old Mongo and Split? (I *think* it will, but I want assurance that you tried it, particularly with child_isinstance()).

antoviaque commented:
@sarina For a test XBlock using this, there is currently only the mentoring XBlock - however it should be relatively straightforward to load it in a current master (birch won't be enough as it is missing patches to XBlock core). For the installation instructions, you should be able to follow https://github.com/open-craft/xblock-mentoring/#installation (or alternatively use https://github.com/open-craft/xblock-mentoring/blob/363a51924eebd7c25a796b6b884fd0baa71b25e1/.travis.yml#L12 if you prefer to install it as a package). If you have any issues with this, let me know, happy to go through this with you.

And agreed that Travis would be useful here - we can add the travis configuration. Who should we ask to enable travis builds for the current repo?

sarina commented:
I think you can ask Ops to enable travis for you on the repo.

I'll test this out over the weekend or early next week. Let me know when my
feedback's been responded to and I'll take another look through.

On Fri, Mar 13, 2015 at 2:45 PM, Xavier Antoviaque <notifications@github.com
> wrote:

> @sarina <https://github.com/sarina> For a test XBlock using this, there
> is currently only the mentoring XBlock - however it should be relatively
> straightforward to load it in a current master (birch won't be enough as it
> is missing patches to XBlock core). For the installation instructions, you
> should be able to follow
> https://github.com/open-craft/xblock-mentoring/#installation (or
> alternatively use
> https://github.com/open-craft/xblock-mentoring/blob/363a51924eebd7c25a796b6b884fd0baa71b25e1/.travis.yml#L12
> if you prefer to install it as a package). If you have any issues with
> this, let me know, happy to go through this with you.
>
> And agreed that Travis would be useful here - we can add the travis
> configuration. Who should we ask to enable travis builds for the current
> repo?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/xblock-utils/pull/5#issuecomment-79239973>.
>

antoviaque commented:
@sarina Sounds good - for testing, also note that you should be able to use the sandbox from the mentoring PR, ie http://sandbox.opencraft.com:18010/

I'll ask DevOps about activating Jenkins here. For addressing the feedback, @bradenmacdonald is back today and that would be the first priority.

bradenmacdonald commented:
@sarina Thanks for the review comments! I have pushed a commit to address all of them (I think) except for a few where I've replied with an explanation.

I like your idea of setting up Travis CI and then rebasing onto that. Will work on that tomorrow.

antoviaque commented:
@bradenmacdonald Thanks! FYI the Travis builds should now be activated on this repo, when a travis.yml file is present. Cf @sarina 's suggestion above to base it off https://github.com/edx/i18n-tools/blob/master/.travis.yml including the linting scripts.

sarina commented on tests/integration/test_studio_editable.py:
```diff
@@ -0,0 +1,301 @@
+            notification = self.dequeue_runtime_notification()
+            self.assertEqual(notification[0], "error")
+            self.assertEqual(notification[1]["title"], "Unable to update settings")
+            self.assertEqual(notification[1]["message"], expected_message)
```
diff-cover asserts this line is uncovered, which seems quite strange if the previous two lines are covered.

sarina replied:
ugh I'm in the wrong file, ignore

sarina commented on xblockutils/studio_editable.py:
```diff
@@ -0,0 +1,320 @@
+        if info["type"] in ("list", "set"):
+            info["value"] = [json.dumps(val) for val in info["value"]]
+            if info["default"] is None:
+                info["default"] = []
```
Diff-cover reports _this_ line - and a few others- are untested.
```
xblockutils.studio_editable     145     29    80%   137, 141, 149, 158-162, 201, 255-257, 272-286, 298-304, 311-313, 320
```

I don't think 100% coverage is required, just want to make sure you've looked over the list of untested lines and made a judgement for each instance as to whether or not it's worth adding an additional test.

sarina commented:
@bradenmacdonald @antoviaque I think this is looking great and we're about ready to merge it (Dave O looked over it last week and is OK with the approach architecturally and didn't have much to add).

Before merging, I'd like responses to:
- Test coverage: are any of the uncovered lines in studio_editable important to write tests for
- Split v mongo: Does this work on both old Mongo and Split? (I think it will, but I want assurance that you tried it, particularly with child_isinstance()).
- On your sandbox, when I [go to add a new question component](http://sandbox.opencraft.com:18010/container/block-v1:TestX+TST-MENTORING-2+now+type@mentoring+block@b10382e331ef4d3eb8824f19d489a7db), I get a message "This component has validation issues" - am I looking at the right test course, and is this normal? It seems to be a scary warning. 

![screen shot 2015-03-20 at 9 47 46 am](https://cloud.githubusercontent.com/assets/1985317/6752606/4c523cca-cee6-11e4-81dc-b1b1f406320a.png)

bradenmacdonald commented:
@sarina 

1. Ok, there are probably a couple more lines I should add tests for. Will do.
2. Yes, we have tested this (via mentoring) on both old mongo and split.
3. Yeah, we debated whether or not to show that warning on new blocks in an earlier PR. It's a warning that the multiple choice question is invalid because you haven't gone in and added any choices yet. Studio hides the exact message until you click "View >" though, so it's a bit hard to tell how important the message is.

bradenmacdonald commented:
I have improved the test coverage. Caught a bug while I was at it, so that was good. The remaining uncovered lines require the Studio environment, so we aren't able to write integration tests for them here.

sarina commented:
@bradenmacdonald ok, that makes sense.

Can you check out the test failures? Once they're passing this is good to merge.

bradenmacdonald commented:
@sarina  Ah, I didn't notice that. Was a simple issue and easy to fix. Merging now. Thanks!
